File: genevent.py
#!/usr/bin/env python3

import re
import time
import json
from datetime import tzinfo, timedelta, datetime, time as dtime
import pypyodbc as pyodbc
# import pyodbc
# import webcolors

db = None
def dbConn(db_host = 'localhost', db_name = 'domjudge', db_user = 'domjudge', db_password = ''):
    connection_string = 'Driver={MySQL ODBC 8.0 Unicode Driver};Server=' + db_host + ';Database=' + db_name + ';UID=' + db_user + ';PWD=' + db_password + ';'
    return pyodbc.connect(connection_string)

cid = 0
st = 0
class UTC(tzinfo):
    def __init__(self,offset = 0):
        self._offset = offset
    def utcoffset(self, dt):
        return timedelta(hours=self._offset)
    def tzname(self, dt):
        return "UTC +%s" % self._offset
    def dst(self, dt):
        return timedelta(hours=self._offset)

def dbGetAll(sheet, nc = False, id = 'id'):
    dic = {}
    cursor = db.cursor()
    where = "" if nc else " WHERE cid = %d" % cid
    cursor.execute("SELECT * FROM " + sheet + where)
    row = cursor.fetchone()
    while row:
        dic[row[id]] = {}
        for i in row.field_dict: dic[row[id]][i] = row[i]
        row = cursor.fetchone()
    cursor.close()
    return dic

def select_contest():
    global cid
    contests = dbGetAll("contest", True, 'cid')
    if len(contests) == 0:
        # raise("No contest found")
        print("No contest found")
        return -1
    if len(contests) == 1:
        for i in contests:
            cid = i
            return contests[i]
    now = time.time()
    print("#\tshortname\tactive\tfinalize\tpublic\tname")
    for i in contests:
        print("%d\t%s\t%s\t%s\t%s\t%s" % (
            i,
            contests[i]["shortname"],
            "Active" if contests[i]["activatetime"] <= now and contests[i]["deactivatetime"] > now and contests[i]["enabled"] else "Deactive",
            "True" if contests[i]["finalizetime"] else "False",
            "True" if contests[i]["public"] else "False",
            contests[i]["name"]
        ))
    cid = int(input("Input the contest id you want to gen feed-event: "))
    if not cid in contests:
        print("Contest not found")
        return -1
    return contests[cid]

def stamp2str(t):
    return datetime.fromtimestamp(t, tz=UTC(8)).isoformat()
def timedura(a, b):
    # return dtime(a - b).isoformat()
    return datetime.fromtimestamp(a - b, tz=UTC(0)).time().isoformat()
# contest
# # state
def gen_contest(info):
    global st
    st = info["starttime"]
    contest = {
        "formal_name": info["name"],
        "penalty_time": 20,
        "start_time": stamp2str(info["starttime"]),
        "end_time":   stamp2str(info["endtime"]),
        "duration":                   timedura(info["endtime"], info["starttime"]),
        "scoreboard_freeze_duration": timedura(info["endtime"], info["freezetime"]),
        "id": cid,
        "external_id": info["externalid"],
        "name": info["name"],
        "shortname": info["shortname"]
    }
    state = {
        "started":        None if not info["starttime"]    else stamp2str(info["starttime"]),
        "ended":          None if not info["endtime"]      else stamp2str(info["endtime"]),
        "frozen":         None if not info["freezetime"]   else stamp2str(info["freezetime"]),
        "thawed":         None if not info["unfreezetime"] else stamp2str(info["unfreezetime"]),
        "finalized":      None if not info["finalizetime"] else stamp2str(info["finalizetime"]),
        "end_of_updates": None
    }
    return { "contest": contest, "state": state }

# judgement-types
def gen_judgement_type():
    dic = {
        "AC":  {"id": "AC",  "name": "correct",            "penalty": False, "solved": True},
        "CE":  {"id": "CE",  "name": "compiler error",     "penalty": False, "solved": False},
        "MLE": {"id": "MLE", "name": "memory limit",       "penalty": True,  "solved": False},
        "NO":  {"id": "NO",  "name": "no output",          "penalty": True,  "solved": False},
        "OLE": {"id": "OLE", "name": "output limit",       "penalty": True,  "solved": False},
        "PE":  {"id": "PE",  "name": "presentation error", "penalty": True,  "solved": False},
        "RTE": {"id": "RTE", "name": "run error",          "penalty": True,  "solved": False},
        "TLE": {"id": "TLE", "name": "timelimit",          "penalty": True,  "solved": False},
        "WA":  {"id": "WA",  "name": "wrong answer",       "penalty": True,  "solved": False}
    }
    return dic
# # language
def gen_language():
    dic = {
        "c":  {"id": "c", "name": "C", "extensions": ["c"], "filter_compiler_files": True, "allow_judge": True, "time_factor": 1.0, "require_entry_point": False, "entry_point_description": None},
        "cpp":  {"id": "cpp", "name": "C++", "extensions": ["cpp", "cc", "cxx", "c++"], "filter_compiler_files": True, "allow_judge": True, "time_factor": 1.0, "require_entry_point": False, "entry_point_description": None},
        "java": {"id": "java", "name": "Java", "extensions": ["java"], "filter_compiler_files": True, "allow_judge": True, "time_factor": 2.0, "require_entry_point": False, "entry_point_description": "Main class"},
        "python3":  {"id": "python3", "name": "Python 3", "extensions": ["py3", "py"], "filter_compiler_files": True, "allow_judge": True, "time_factor": 1.0, "require_entry_point": False, "entry_point_description": "Main file"}
    }
    return dic

# testcase
def testcase_count(probid):
    cursor = db.cursor()
    cursor.execute("SELECT COUNT(*) AS c FROM testcase WHERE probid = %d" % probid)
    row = cursor.fetchone()
    count = row["c"]
    cursor.close()
    return count

# problem
# contestproblem
def gen_problem():
    dic = {}
    contestproblems = dbGetAll("contestproblem", id='probid')
    problems = dbGetAll("problem", True, 'probid')
    c = 0
    for i in contestproblems:
        item = {
            "ordinal": c,
            "id": i,
            "short_name": contestproblems[i]["shortname"],
            "label": contestproblems[i]["shortname"],
            "time_limit": problems[i]["timelimit"],
            "externalid": problems[i]["externalid"],
            "name": problems[i]["name"],
            "rgb": contestproblems[i]["color"], # name_to_hex(contestproblems[i]["color"])
            "color": contestproblems[i]["color"],
            "test_data_count": testcase_count(i)
        }
        c += 1
        dic[i] = item
    return dic

# groups
# # team_category
def gen_group():
    dic = {}
    categorys = dbGetAll("team_category", True, 'categoryid')
    for i in categorys:
        item = {
            "hidden": not bool(categorys[i]["visible"]),
            "icpc_id": categorys[i]["categoryid"], # Same as id
            "id": i,
            "name": categorys[i]["name"],
            "sortorder": categorys[i]["sortorder"],
            "color": categorys[i]["color"]
        }
        dic[i] = item
    return dic

# organizations
# # team_affiliation
def gen_organizations():
    dic = {}
    affiliations = dbGetAll("team_affiliation", True, 'affilid')
    for i in affiliations:
        item = {
            "icpc_id": affiliations[i]["externalid"],
            "shortname": affiliations[i]["shortname"],
            "id": i,
            "name": affiliations[i]["shortname"],
            "formal_name": affiliations[i]["name"],
            "country": affiliations[i]["country"]
        }
        dic[i] = item
    return dic

# team
def gen_team(affiliations):
    dic = {}
    teams = dbGetAll("team", True, 'teamid')
    # affiliations = dbGetAll("team_affiliation", True, 'affilid')
    for i in teams:
        affil = None if not teams[i]["affilid"] else affiliations[teams[i]["affilid"]]
        item = {
            "externalid": teams[i]["externalid"],
            "group_ids": [ teams[i]["categoryid"] ],
            "affiliation": None if not affil else affil["formal_name"],
            "id": i,
            "icpc_id": teams[i]["teamid"], # Same as id
            "name": teams[i]["name"],
            "organization_id": teams[i]["affilid"],
            "members": teams[i]["members"]
        }
        dic[i] = item
    return dic

# submission
def gen_submission():
    dic = {}
    submissions = dbGetAll("submission", id='submitid')
    for i in submissions:
        item = {
            "language_id": submissions[i]["langid"],
            "time": stamp2str(submissions[i]["submittime"]),
            "contest_time": timedura(submissions[i]["submittime"], st),
            "id": i,
            "externalid": submissions[i]["externalid"],
            "team_id": submissions[i]["teamid"],
            "problem_id": submissions[i]["probid"],
            "entry_point": submissions[i]["entry_point"],
            "files":[{
                "href": "contests/%d/submissions/%d/files" % (cid, i),
                "mime": "application/zip"
            }]
        }
        dic[i] = item
    return dic

# # judging_run
# def gen_runs(cid):
# # # judging_run_output

# judging / judgements
def gen_judging(judgement):
    dic = {}
    judgings = dbGetAll("judging", id='judgingid')
    for i in judgings:
        if not judgings[i]["valid"]: continue
        item = {
            "max_run_time": float(0),
            "start_time": stamp2str(judgings[i]["starttime"]),
            "start_contest_time": timedura(judgings[i]["starttime"], st),
            "end_time": stamp2str(judgings[i]["endtime"]),
            "end_contest_time": timedura(judgings[i]["endtime"], st),
            "id": i,
            "submission_id": judgings[i]["submitid"],
            "valid": bool(judgings[i]["valid"]),
            "judgehost": judgings[i]["judgehost"],
            "judgement_type_id": judgement[judgings[i]["result"]]
        }
        dic[i] = item
    return dic

# # clarifications

static_event_id = 1
def genEvent(typ, data, id=0, op="create", time=None):
    global static_event_id
    e = {
        "id": id if id else static_event_id,
        "type": typ,
        "op": op,
        "data": data,
        "time": time if time else datetime.now(UTC(8)).isoformat()
    }
    static_event_id += 1
    return json.dumps(e) + "\n"

def main():
    global db
    db = dbConn(db_user="root")
    contest = select_contest()
    if contest == -1: return
    info = gen_contest(contest)
    judgement_types = gen_judgement_type()
    problem = gen_problem()
    groups = gen_group()
    organizations = gen_organizations()
    teams = gen_team(organizations)
    submissions = gen_submission()
    judgement = {}
    for i in judgement_types: judgement[judgement_types[i]["name"].replace(' ', '-')] = i
    judgements = gen_judging(judgement)
    db.close()

    with open("event-feed.json", 'w') as f:
        f.writelines(genEvent("contests", info["contest"]))
        for i in judgement_types:   f.write(genEvent("judgement-types",judgement_types[i]))
        for i in problem:           f.write(genEvent("problems",       problem[i]))
        for i in groups:            f.write(genEvent("groups",         groups[i]))
        for i in organizations:     f.write(genEvent("organizations",  organizations[i]))
        for i in teams:             f.write(genEvent("teams",          teams[i]))
        # f.writelines(genEvent("state", info["state"], op="create"))
        for i in submissions:       f.write(genEvent("submissions",    submissions[i]))
        for i in judgements:        f.write(genEvent("judgements",     judgements[i], op="create"))
        # for i in judgements:        f.write(genEvent("judgements",     judgements[i], op="update"))
        f.writelines(genEvent("state", info["state"], op="update"))

    with open("event.json", 'w') as f:
        data = {
            "contest": {
                "info": info["contest"],
                "state": info["state"],
                "problem":      list(problem.values()),
                "groups":       list(groups.values()),
                "organizations":list(organizations.values()),
                "teams":        list(teams.values()),
                "submissions":  list(submissions.values()),
                "judgements":   list(judgements.values()),
            }
        }
        json.dump(data, f, ensure_ascii=False, indent=2, separators=(',', ': '))

if __name__ == "__main__":
    main()

File: genevent.php
#!/usr/bin/env php
<?php
$con = null;
$cid = 0;
$st = 0;
function dbConn($db_host = 'localhost', $db_name = 'domjudge', $db_user = 'domjudge', $db_password = ''){
    $con = new mysqli($db_host, $db_user, $db_password, $db_name);
    if (!$con) {
        die('Could not connect: ' . mysqli_error());
    }
    return $con;
}

function dbGetAll($sheet, $nc = False, $id = 'id') {
    global $con, $cid;
    $dic = [];
    $where = $nc ? "" : sprintf(" WHERE cid = %d", $cid);
    $cursor = $con->query("SELECT * FROM " . $sheet . $where);
    while($row = $cursor->fetch_assoc()) { //->fetch_array(MYSQLI_ASSOC)
        $dic[$row[$id]] = $row;
    }
    $cursor->close();
    return $dic;
}

function testcase_count($probid) {
    global $con;
    $cursor = $con->query(sprintf("SELECT COUNT(*) AS c FROM testcase WHERE probid = %d", $probid));
    $row = $cursor->fetch_array(MYSQLI_NUM);
    $count = $row[0];
    $cursor->close();
    return $count;
}

function stamp2str($t) {
    return date(DATE_ISO8601, $t);
}

function timedura($a, $b) {
    return gmstrftime('%H:%M:%S', $a - $b);
}

function select_contest() {
    global $cid;
    $contests = dbGetAll("contest", True, 'cid');
    if (count($contests) == 0) {
        error_log("No contest found");
        return -1;
    }
    if (count($contests) == 1) {
        foreach($contests as $i=>$v){
            $cid = $i;
            return $v;
        }
    }
    $now = time();
    if (isset($_GET['cid'])) {
        $cid = (int) $_GET['cid'];
        if (!array_key_exists($cid, $contests)) return -1;
    } else {
        error_log("#\tshortname\tactive\tfinalize\tpublic\tname");
        foreach($contests as $i=>$v){
            error_log(sprintf("%d\t%s\t%s\t%s\t%s\t%s",
                $i,
                $v["shortname"],
                $v["activatetime"] <= $now and $v["deactivatetime"] > $now and $v["enabled"],
                $v["finalizetime"] ? true : false,
                $v["public"] ? true : false,
                $v["name"]
            ));
        }
        sscanf(fgets(STDIN), "%d", $cid);
        if (!array_key_exists($cid, $contests)) return -1;
    }
    return $contests[$cid];
}

function gen_contest($info) {
    global $cid, $st;
    $st = $info["starttime"];
    $contest = [
        "formal_name" => $info["name"],
        "penalty_time" => 20,
        "start_time" => stamp2str($info["starttime"]),
        "end_time" =>   stamp2str($info["endtime"]),
        "duration" =>                   timedura($info["endtime"], $info["starttime"]),
        "scoreboard_freeze_duration" => timedura($info["endtime"], $info["freezetime"]),
        "id" => $cid,
        "external_id" => $info["externalid"],
        "name" => $info["name"],
        "shortname" => $info["shortname"]
    ];
    $state = [
        "started" =>        $info["starttime"]    ? stamp2str($info["starttime"]   ) : null,
        "ended" =>          $info["endtime"]      ? stamp2str($info["endtime"]     ) : null,
        "frozen" =>         $info["freezetime"]   ? stamp2str($info["freezetime"]  ) : null,
        "thawed" =>         $info["unfreezetime"] ? stamp2str($info["unfreezetime"]) : null,
        "finalized" =>      $info["finalizetime"] ? stamp2str($info["finalizetime"]) : null,
        "end_of_updates" => null
    ];
    return [ "contest" => $contest, "state" => $state ];
}

function gen_judgement_type() {
    $dic = [
        "AC"  => ["id" => "AC",  "name" => "correct",            "penalty" => False, "solved" => True ],
        "CE"  => ["id" => "CE",  "name" => "compiler error",     "penalty" => False, "solved" => False],
        "MLE" => ["id" => "MLE", "name" => "memory limit",       "penalty" => True,  "solved" => False],
        "NO"  => ["id" => "NO",  "name" => "no output",          "penalty" => True,  "solved" => False],
        "OLE" => ["id" => "OLE", "name" => "output limit",       "penalty" => True,  "solved" => False],
        "PE"  => ["id" => "PE",  "name" => "presentation error", "penalty" => True,  "solved" => False],
        "RTE" => ["id" => "RTE", "name" => "run error",          "penalty" => True,  "solved" => False],
        "TLE" => ["id" => "TLE", "name" => "timelimit",          "penalty" => True,  "solved" => False],
        "WA"  => ["id" => "WA",  "name" => "wrong answer",       "penalty" => True,  "solved" => False]
    ];
    return $dic;
}

function gen_language() {
    $dic = [
        "c"       => [
            "id"                      => "c",
            "name"                    => "C",
            "extensions"              => ["c"],
            "filter_compiler_files"   => True,
            "allow_judge"             => True,
            "time_factor"             => 1.0,
            "require_entry_point"     => False,
            "entry_point_description" => null
        ],
        "cpp"     => [
            "id"                      => "cpp",
            "name"                    => "C++",
            "extensions"              => ["cpp", "cc", "cxx", "c++"],
            "filter_compiler_files"   => True,
            "allow_judge"             => True,
            "time_factor"             => 1.0,
            "require_entry_point"     => False,
            "entry_point_description" => null
        ],
        "java"    => [
            "id"                      => "java",
            "name"                    => "Java",
            "extensions"              => ["java"],
            "filter_compiler_files"   => True,
            "allow_judge"             => True,
            "time_factor"             => 2.0,
            "require_entry_point"     => False,
            "entry_point_description" => "Main class"
        ],
        "python3" => [
            "id"                      => "python3",
            "name"                    => "Python 3",
            "extensions"              => ["py3", "py"],
            "filter_compiler_files"   => True,
            "allow_judge"             => True,
            "time_factor"             => 1.0,
            "require_entry_point"     => False,
            "entry_point_description" => "Main file"
        ]
    ];
    return $dic;
}

function gen_problem() {
    $dic = [];
    $contestproblems = dbGetAll("contestproblem", False, 'probid');
    $problems = dbGetAll("problem", True, 'probid');
    $c = 0;
    foreach($contestproblems as $i=>$v) {
        $item = [
            "ordinal"         => $c,
            "id"              => $i,
            "short_name"      => $v["shortname"],
            "label"           => $v["shortname"],
            "time_limit"      => $problems[$i]["timelimit"],
            "externalid"      => $problems[$i]["externalid"],
            "name"            => $problems[$i]["name"],
            "rgb"             => $v["color"],
            "color"           => $v["color"],
            "test_data_count" => testcase_count($i)
        ];
        $c += 1;
        $dic[$i] = $item;
    }
    return $dic;
}

function gen_group() {
    $dic = [];
    $categorys = dbGetAll("team_category", True, 'categoryid');
    foreach($categorys as $i=>$v) {
        $item = [
            "hidden"    => !$v["visible"],
            "icpc_id"   => $v["categoryid"], # Same as id
            "id"        => $i,
            "name"      => $v["name"],
            "sortorder" => $v["sortorder"],
            "color"     => $v["color"]
        ];
        $dic[$i] = $item;
    }
    return $dic;
}

function gen_organizations() {
    $dic = [];
    $affiliations = dbGetAll("team_affiliation", True, 'affilid');
    foreach($affiliations as $i=>$v) {
        $item = [
            "icpc_id"     => $v["externalid"],
            "shortname"   => $v["shortname"],
            "id"          => $i,
            "name"        => $v["shortname"],
            "formal_name" => $v["name"],
            "country"     => $v["country"]
        ];
        $dic[$i] = $item;
    }
    return $dic;
}

function gen_team($affiliations) {
    $dic = [];
    $teams = dbGetAll("team", True, 'teamid');
    foreach($teams as $i=>$v) {
        $affil = isset($v["affilid"]) ? $affiliations[$v["affilid"]] : null;
        $item = [
            "externalid"      => $v["externalid"],
            "group_ids"       => [ $v["categoryid"] ],
            "affiliation"     => isset($affil) ? $affil["formal_name"] : null,
            "id"              => $i,
            "icpc_id"         => $v["teamid"], # Same as id
            "name"            => $v["name"],
            "organization_id" => $v["affilid"],
            "members"         => $v["members"]
        ];
        $dic[$i] = $item;
    }
    return $dic;
}

function gen_submission() {
    global $cid, $st;
    $dic = [];
    $submissions = dbGetAll("submission", False, 'submitid');
    foreach($submissions as $i=>$v) {
        $item = [
            "language_id"  => $v["langid"],
            "time"         => stamp2str($v["submittime"]),
            "contest_time" => timedura($v["submittime"], $st),
            "id"           => $i,
            "externalid"   => $v["externalid"],
            "team_id"      => $v["teamid"],
            "problem_id"   => $v["probid"],
            "entry_point"  => $v["entry_point"],
            "files"        => [[
                "href" => sprintf("contests/%d/submissions/%d/files", $cid, $i),
                "mime" => "application/zip"
            ]]
        ];
        $dic[$i] = $item;
    }
    return $dic;
}

# # judging_run
# function gen_runs() {}
# # # judging_run_output

function gen_judging($judgement) {
    global $st;
    $dic = [];
    $judgings = dbGetAll("judging", False, 'judgingid');
    foreach($judgings as $i=>$v) {
        $item = [
            "max_run_time"       => (float) 0,
            "start_time"         => stamp2str($v["starttime"]),
            "start_contest_time" => timedura($v["starttime"], $st),
            "end_time"           => stamp2str($v["endtime"]),
            "end_contest_time"   => timedura($v["endtime"], $st),
            "id"                 => $i,
            "submission_id"      => $v["submitid"],
            "valid"              => (bool) $v["valid"],
            "judgehost"          => $v["judgehost"],
            "judgement_type_id"  => $judgement[$v["result"]] ?? $v["result"]
        ];
        $dic[$i] = $item;
    }
    return $dic;
}
# # clarifications

function genEvent($typ, $data, $id=null, $op="create", $time=null) {
    static $event_id = 1;
    $e = [
        "id"   => $id ?? $event_id,
        "type" => $typ,
        "op"   => $op,
        "data" => $data,
        "time" => $time ?? date(DATE_ISO8601)
    ];
    $event_id += 1;
    return json_encode($e) . "\n";
}

function main() {
    global $con;
    $con = dbConn('localhost', 'domjudge', 'root');
    $contest = select_contest();
    if ($contest == -1) return;
    $info = gen_contest($contest);
    $judgement_types = gen_judgement_type();
    $problem = gen_problem();
    $groups = gen_group();
    $organizations = gen_organizations();
    $teams = gen_team($organizations);
    $submissions = gen_submission();
    $judgement = [];
    foreach ($judgement_types as $i=>$v) {
        $judgement[str_replace(' ', '-', $v["name"])] = $i;
    }
    $judgements = gen_judging($judgement);
    mysqli_close($con);

    echo genEvent("contests", $info["contest"]);
    foreach($judgement_types as $i=>$v) echo genEvent("judgement-types", $v);
    foreach($problem         as $i=>$v) echo genEvent("problem",         $v);
    foreach($groups          as $i=>$v) echo genEvent("groups",          $v);
    foreach($organizations   as $i=>$v) echo genEvent("organizations",   $v);
    foreach($teams           as $i=>$v) echo genEvent("teams",           $v);
    foreach($submissions     as $i=>$v) echo genEvent("submissions",     $v);
    foreach($judgements      as $i=>$v) echo genEvent("judgements",      $v);
    echo genEvent("state", $info["state"]);
    /*
    $f = fopen("event.json", 'w');
    $data = [
        "contest" => [
            "info"          => $info["contest"],
            "state"         => $info["state"],
            "problem"       => array_values($problem),
            "groups"        => array_values($groups),
            "organizations" => array_values($organizations),
            "teams"         => array_values($teams),
            "submissions"   => array_values($submissions),
            "judgements"    => array_values($judgements),
        ]
    ];
    fwrite($f, json_encode($data));
    fclose($f);
    */
}

main();
?>